### PR TITLE
Codes for preprocessing

### PR DESCRIPTION
**Pull request by samwang141224**
DesiMock.py defines a class DesiMock to save data
preprocess.py is about the preprocessing, such as rebin, normalize and estimate SNR

### CODE REVIEW

File: desidlas/datasets/DesiMock.py [hunk 1]
```diff
@@ -0,0 +1,155 @@
...
+            """
+            
+            Parameters
+            ----------
+            start_point: int, the start index of the slice of the data(wavelength, flux, error), default 0
+            end_point: int, the end index of the slice of the data(wavelength, flux, error), default the length of the data array
+            
+            Returns
+            -------
+
```
Comment (profxj) [inline desidlas/datasets/DesiMock.py:111]: Describe the `sightline` that is returned

File: desidlas/datasets/init.py [hunk 2]
```diff
@@ -0,0 +1 @@
+
```
Comment (profxj) [inline desidlas/datasets/init.py]: should this be `__init__.py` ?

File: desidlas/datasets/preprocess.py [hunk 3]
```diff
@@ -0,0 +1,325 @@
...
+
+def label_sightline(sightline, kernel=kernel, REST_RANGE=REST_RANGE, pos_sample_kernel_percent=0.3):
+    """
+    Add labels to input sightline based on the DLAs along that sightline
+
+    Parameters
+    ----------
+    sightline: dla_cnn.data_model.Sightline
+    pos_sample_kernel_percent: float
+    kernel: int
```
Comment (profxj) [inline desidlas/datasets/preprocess.py]: describe `kernel` and `REST_RANGE`

File: desidlas/datasets/preprocess.py [hunk 4]
```diff
@@ -0,0 +1,325 @@
...
+
+    # Append these to the Sightline
+    sightline.classification = classification
+    sightline.offsets = offsets_array
+    sightline.column_density = column_density
+
+    # classification is 1 / 0 / -1 for DLA/nonDLA/border
+    # offsets_array is offset
+    return classification, offsets_array, column_density
+def rebin(sightline, v):
```
Comment (profxj) [inline desidlas/datasets/preprocess.py:104]: add blank line

File: desidlas/datasets/preprocess.py [hunk 5]
```diff
@@ -0,0 +1,325 @@
...
+    Resample and rebin the input Sightline object's data to a constant dlambda/lambda dispersion.
+    Parameters
+    ----------
+    sightline: :class:`dla_cnn.data_model.Sightline.Sightline`
+    v: float, and np.log(1+v/c) is dlambda/lambda, its unit is m/s, c is the velocity of light
+    Returns
+    -------
+    :class:`dla_cnn.data_model.Sightline.Sightline`:
+    """
+    # TODO -- Add inline comments
```
Comment (profxj) [inline desidlas/datasets/preprocess.py:115]: Do this

File: desidlas/datasets/preprocess.py [hunk 6]
```diff
@@ -0,0 +1,325 @@
...
+
+
+def normalize(sightline, full_wavelength, full_flux):
+    """
+    Normalize spectrum by dividing the mean value of continnum at lambda[left,right]
+    ------------------------------------------
+    parameters:
+    
+    sightline: dla_cnn.data_model.Sightline.Sightline object;
+    camera : str, 'b' : the blue channel of the specctra, 'r': the r channel of the spectra,
```
Comment (profxj) [inline desidlas/datasets/preprocess.py]: `camera` is not a parameter

File: desidlas/datasets/preprocess.py [hunk 7]
```diff
@@ -0,0 +1,325 @@
...
+            for dla in sightline.dlas:
+                dlas_col_density += str(dla.col_density)+","
+                dlas_central_wavelength += str(dla.central_wavelength)+","
+            info["dlas_col_density"] = dlas_col_density[:-1]
+            info["dlas_central_wavelength"] = dlas_central_wavelength[:-1]
+            
+            #write to the csv file
+            summary_table_writer.writerow(info)
+#from dla_cnn.desi.DesiMock import DesiMock
+def write_summary_table(nums, version,path, output_path):
```
Comment (profxj) [inline desidlas/datasets/preprocess.py:290]: add blank line

File: desidlas/datasets/preprocess.py [hunk 8]
```diff
@@ -0,0 +1,325 @@
...
+    """
+    Directly read data from fits files and write the summary table, the summary table contains all available sightlines(dlas!=[] and z_qso>2.33) in the given fits files.
+    -----------------------------------------------------------------------------------------------------------------------------------------
+    parameters:
+    nums: list, the given fits files' id, its elements' format is int, and one should make sure all fits files are available before invoking this funciton, otherwise some sightlines can be missed;
+    version: int, the version of the data set we use, e.g. if the version is v9.16, then version = 16
+    path: str, the dir of the folder which stores the given fits file, the folder's structure is like folder-fits files' id - fits files , if you are still confused, you can check the below code about read data from the fits file;
+    output_path: str, the dir where the summary table is generated, and if there have been a summary table, then we will remove it and generate a new summary table;
+    ------------------------------------------------------------------------------------------------------------------------------------------
+    retrun:
```
Comment (profxj) [inline desidlas/datasets/preprocess.py]: typo

**Review by profxj — CHANGES_REQUESTED** [review]
Very good!

Please add several unit tests that touch this code.
Let me know if you need help doing so.

File: desidlas/datasets/DesiMock.py (hunk 1 above)
Comment (samwang141224) [reply desidlas/datasets/DesiMock.py:111]: Done

File: desidlas/datasets/init.py (hunk 2 above)
Comment (samwang141224) [reply desidlas/datasets/init.py]: I have corrected it

File: desidlas/datasets/preprocess.py (hunk 3 above)
Comment (samwang141224) [reply desidlas/datasets/preprocess.py]: Done

File: desidlas/datasets/preprocess.py (hunk 4 above)
Comment (samwang141224) [reply desidlas/datasets/preprocess.py:104]: Done

File: desidlas/datasets/preprocess.py (hunk 6 above)
Comment (samwang141224) [reply desidlas/datasets/preprocess.py]: I have corrected it

File: desidlas/datasets/preprocess.py (hunk 7 above)
Comment (samwang141224) [reply desidlas/datasets/preprocess.py:290]: Done

File: desidlas/datasets/preprocess.py (hunk 8 above)
Comment (samwang141224) [reply desidlas/datasets/preprocess.py]: This function has no returns, so should I just delete the return

File: desidlas/datasets/preprocess.py (hunk 5 above)
Comment (samwang141224) [reply desidlas/datasets/preprocess.py:115]: Done

File: desidlas/datasets/preprocess.py (hunk 8 above)
Comment (profxj) [reply desidlas/datasets/preprocess.py]: yes, please do

Comment (samwang141224) [reply desidlas/datasets/preprocess.py]: Done

File: desidlas/tests/test_preprocess.py [hunk 9]
```diff
@@ -0,0 +1,54 @@
...
+
+datafile_path = os.path.join(resource_filename('desidlas', 'tests'), 'datafile')
+
+#load spectra file for testing
+spectra= os.path.join(datafile_path, 'spectra', 'spectra-16-1375.fits')
+truth=os.path.join(datafile_path, 'spectra', 'truth-16-1375.fits')
+zbest=os.path.join(datafile_path, 'spectra', 'zbest-16-1375.fits')
+
+#read spectra data using DesiMock
+specs = DesiMock()
```
Comment (profxj) [inline desidlas/tests/test_preprocess.py]: Put this code in test methods.  And then do a test, e.g.
an `assert` command.

File: desidlas/tests/test_preprocess.py [hunk 10]
```diff
@@ -0,0 +1,54 @@
...
+#get sightline id in the spectra file
+keys = list(specs.data.keys())
+jj=keys[0] #choose 1 sightline to do the test
+
+#m
+sightline = specs.get_sightline(jj,camera = 'all', rebin=False, normalize=True)
+
+def test_estimate_s2n(sightline):
+    # estimate the signal-to-noise ratio
+    sightline.s2n=estimate_s2n(sightline)
```
Comment (profxj) [inline desidlas/tests/test_preprocess.py:53]: Add a test, e.g. an `assert` statement that checks the `s2n` value.

File: desidlas/tests/test_preprocess.py [hunk 11]
```diff
@@ -0,0 +1,54 @@
...
+    
+
+def test_label_sightline():
```
Comment (profxj) [inline desidlas/tests/test_preprocess.py]: add tests here or remove these

**Review by profxj — CHANGES_REQUESTED** [review]
just a few final requests

File: desidlas/tests/test_preprocess.py (hunk 9 above)
Comment (samwang141224) [reply desidlas/tests/test_preprocess.py]: I have added three asserts:
1, check if the data file exists
2, check if there is at least one spectra in the DesiMock
3, check if the SNR value is above 0

File: desidlas/tests/test_preprocess.py [hunk 12]
```diff
@@ -0,0 +1,56 @@
...
+assert Path(truth).exists(), 'no truth file! please check your data'
+assert Path(zbest).exists(), 'no zbest file! please check your data'
+
+#read spectra data using DesiMock
+specs = DesiMock()
+specs.read_fits_file(spectra,truth,zbest)
+
+#get sightline id in the spectra file
+keys = list(specs.data.keys())
+assert len(keys)>0, 'no spectra in the data file'
```
Comment (profxj) [inline desidlas/tests/test_preprocess.py]: Put these lines inside a test, e.g.

def test_io():
    code to load file
    assert len(keys) > 0

**Review by profxj — APPROVED** [review]
approved!

**Pull request merged**